### PR TITLE
[SPARK-29869][SQL] improve error message in HiveMetastoreCatalog#convertToLogicalRelation 

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?
In our production, HiveMetastoreCatalog#convertToLogicalRelation throws AssertError occasionally:
```sql
scala> spark.table("hive_table").show
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:208)
  at org.apache.spark.sql.hive.HiveMetastoreCatalog.convertToLogicalRelation(HiveMetastoreCatalog.scala:261)
  at org.apache.spark.sql.hive.HiveMetastoreCatalog.convert(HiveMetastoreCatalog.scala:137)
  at org.apache.spark.sql.hive.RelationConversions$$anonfun$apply$4.applyOrElse(HiveStrategies.scala:220)
  at org.apache.spark.sql.hive.RelationConversions$$anonfun$apply$4.applyOrElse(HiveStrategies.scala:207)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsDown$2(AnalysisHelper.scala:108)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:72)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsDown$1(AnalysisHelper.scala:108)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsDown(AnalysisHelper.scala:106)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsDown$(AnalysisHelper.scala:104)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsDown(LogicalPlan.scala:29)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsDown$4(AnalysisHelper.scala:113)
  at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$mapChildren$1(TreeNode.scala:376)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:214)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:374)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:327)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsDown$1(AnalysisHelper.scala:113)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsDown(AnalysisHelper.scala:106)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsDown$(AnalysisHelper.scala:104)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsDown(LogicalPlan.scala:29)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperators(AnalysisHelper.scala:73)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperators$(AnalysisHelper.scala:72)
  at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperators(LogicalPlan.scala:29)
  at org.apache.spark.sql.hive.RelationConversions.apply(HiveStrategies.scala:207)
  at org.apache.spark.sql.hive.RelationConversions.apply(HiveStrategies.scala:191)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$2(RuleExecutor.scala:130)
  at scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:60)
  at scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:68)
  at scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:49)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1(RuleExecutor.scala:127)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1$adapted(RuleExecutor.scala:119)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:119)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.org$apache$spark$sql$catalyst$analysis$Analyzer$$executeSameContext(Analyzer.scala:168)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.execute(Analyzer.scala:162)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.execute(Analyzer.scala:122)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$executeAndTrack$1(RuleExecutor.scala:98)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker$.withTracker(QueryPlanningTracker.scala:88)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.executeAndTrack(RuleExecutor.scala:98)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$executeAndCheck$1(Analyzer.scala:146)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.markInAnalyzer(AnalysisHelper.scala:201)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.executeAndCheck(Analyzer.scala:145)
  at org.apache.spark.sql.execution.QueryExecution.$anonfun$analyzed$1(QueryExecution.scala:66)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
  at org.apache.spark.sql.execution.QueryExecution.analyzed$lzycompute(QueryExecution.scala:63)
  at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:63)
  at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:55)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:86)
  at org.apache.spark.sql.SparkSession.table(SparkSession.scala:585)
  at org.apache.spark.sql.SparkSession.table(SparkSession.scala:581)
  ... 47 elided
````
Most of cases occurred in reading a table which created by an old Spark version.
After recreated the table, the issue will be gone.

After deep dive, the root cause is this external table is a non-partitioned table but the `LOCATION` set to a partitioned path {{/tablename/dt=yyyymmdd}}. The partitionSpec is inferred.

### Why are the changes needed?
Above error message is very confused. We need more details about assert failure information.

This issue caused by `PartitioningAwareFileIndex#inferPartitioning()`. For non-HiveMetastore Spark, it's useful. But for Hive table, it shouldn't infer partition if Hive tell us it's a non partitioned table. (new added)

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Add UT.

### CODE REVIEW

**Comment by LantaoJin** [comment]
cc @wangyum , more input?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -257,8 +257,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+    assert(result.output.length == relation.output.length,
+      s"Target table has ${result.output.length} columns, " +
+        s"but source table has ${relation.output.length} columns. " +
+        s"It may need to recreate the table ${relation.tableMeta.identifier}")
+    result.output.zip(relation.output).foreach { case (a1, a2) =>
+      assert(a1.dataType.sameType(a2.dataType),
+        s"Data type of column ${a1.name} in target table is ${a1.dataType.typeName}, " +
+          s"but column ${a2.name} in source table is ${a2.dataType.typeName}. " +
+          s"It may need to recreate the table ${relation.tableMeta.identifier}")
+    }
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Thank you @LantaoJin Could we add a test case?

Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: `spark.sql.sources.*` can not add in SQL of `CREATE TABLE` in current Spark version. Any other way to go to this invalid code path?

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: `sparkSession.metadataHive.runSqlHive`?
https://github.com/apache/spark/blob/96179732aad28418d486de1a365cd5d68c3db910/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala#L314-L320

**Comment by LantaoJin** [comment]
Hi @wangyum , do you think we should block the creation of non-partitioned table which LOCATION/path contains "=".

**Comment by LantaoJin** [comment]
> Hi @wangyum , do you think should we block the creation of non-partitioned table which LOCATION/path contains "=".

But we cannot ban the creation by Hive.

**Comment by LantaoJin** [comment]
This issue caused by `PartitioningAwareFileIndex#inferPartitioning()`. For non-HiveMetastore Spark (maybe databricks runtime), it's useful. But for Hive table, it shouldn't infer partition if Hive tell us it's a non partitioned table. I think it should be fix by adding a variable in `DataSource#resolveRelation`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 2]
```diff
@@ -362,29 +362,20 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
     withTempPath(dir => {
       val baseDir = s"${dir.getCanonicalFile.toURI.toString}/test"
-      val partitionDir = s"${dir.getCanonicalFile.toURI.toString}/test/dt=20191113"
-      val file = new Path(partitionDir, "file.parquet")
-      val fs = file.getFileSystem(new Configuration())
-      fs.createNewFile(file)
-      withTable("test") {
-        withSQLConf(HiveUtils.CONVERT_METASTORE_PARQUET.key -> "true",
-          SQLConf.PARALLEL_PARTITION_DISCOVERY_THRESHOLD.key -> "0") {
+      val partitionLikeDir = s"${dir.getCanonicalFile.toURI.toString}/test/dt=20191113"
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: `val partitionLikeDir = s"${dir.getCanonicalFile.toURI.toString}/test/dt=20191113"` -> `val partitionLikeDir = s"$baseDir/dt=20191113"`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 3]
```diff
@@ -362,29 +362,20 @@ class DataSourceWithHiveMetastoreCatalogSuite
     }
   }
 
-  test("SPARK-29869: HiveMetastoreCatalog#convertToLogicalRelation throws AssertionError") {
+  test("SPARK-29869: fix HiveMetastoreCatalog#convertToLogicalRelation throws AssertionError") {
     withTempPath(dir => {
       val baseDir = s"${dir.getCanonicalFile.toURI.toString}/test"
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: `test ` -> `non_partition_table `?

**Comment by LantaoJin** [comment]
Soft ping @cloud-fan @HyukjinKwon @wangyum @viirya

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 4]
```diff
@@ -362,29 +362,20 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
-               |OUTPUTFORMAT
-               |  'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat'
-               |LOCATION '$baseDir'
+               |CREATE TABLE non_partition_table (id bigint)
+               |STORED AS PARQUET LOCATION '$baseDir'
                |""".stripMargin)
-          val e = intercept[AssertionError](spark.sql("select * from test")).getMessage
-          assert(e.contains("assertion failed"))
+          assert(spark.sql("select * from non_partition_table").collect() ===
+            Array(Row(0), Row(1), Row(2)))
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: `assert(spark.sql("select * from non_partition_table").collect() === Array(Row(0), Row(1), Row(2)))` -> `checkAnswer(spark.table("non_partition_table"), Row(0, 1, 2))`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala (hunk 3 above)
Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: the location can be set to anywhere. `test` here is fine. `non_partition_table ` will over 100 chars :)

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala (hunk 4 above)
Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: ok. I wanna wait more comments to update in one.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 5]
```diff
@@ -358,4 +361,23 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
+  test("SPARK-29869: fix HiveMetastoreCatalog#convertToLogicalRelation throws AssertionError") {
+    withTempPath(dir => {
+      val baseDir = s"${dir.getCanonicalFile.toURI.toString}/test"
+      val partitionLikeDir = s"${dir.getCanonicalFile.toURI.toString}/test/dt=20191113"
+      spark.range(3).selectExpr("id").write.parquet(partitionLikeDir)
+      withTable("non_partition_table") {
+        withSQLConf(HiveUtils.CONVERT_METASTORE_PARQUET.key -> "true") {
+          spark.sql(
+            s"""
+               |CREATE TABLE non_partition_table (id bigint)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Isn't it a malformed table? Does hive ignore the directories for non-partitioned tables?

Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Seems Hive return none when query this table(1.2.1):
> hive> select * from `xxxxx`.`xxxx`;
OK
Time taken: 25.301 seconds

But no assertion error

Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Maybe you are right. Actually the data file path of this table is `/path/tablename/dt=yyyymmdd/dt=yyyymmdd/xxx.parquet`. I guess Hive does not recursively lookup load the data. So it return empty but not error.
And I found if when enable recursively lookup by `.option("recursiveFileLookup", true)`, the inferPartitioning will be disable. So `dt=yyyymmdd` won't be treated as partitionSpec.

So should I revert the code changes and only keep the assert detail information? Or auto rollback to do not use built-in Parquet reader to read?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 6]
```diff
@@ -245,20 +245,29 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
                 userSpecifiedSchema = Option(updatedTable.dataSchema),
                 bucketSpec = None,
                 options = options,
-                className = fileType).resolveRelation(),
+                className = fileType).resolveRelation(needPartitionInferring = false),
               table = updatedTable)
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Can we just trim inferred partition spec here?

Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Do you mean trimmed here like this?
```scala
          val updatedTable = inferIfNeeded(relation, options, fileFormat)
          val baseRelation = DataSource(
            sparkSession = sparkSession,
            paths = rootPath.toString :: Nil,
            userSpecifiedSchema = Option(updatedTable.dataSchema),
            bucketSpec = None,
            options = options,
            className = fileType).resolveRelation()
          // filter out the inferred partition col
          val trimmedOutput = StructType(baseRelation.schema.filter { field =>
            updatedTable.schema.names.contains(field.name)
          }).toAttributes
          val created = LogicalRelation(baseRelation, trimmedOutput, table = updatedTable)
```
But `baseRelation`.schema is still contains partition field.

Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: `BaseRelation` doesn't have a copy method.

**Comment by LantaoJin** [comment]
Since this issue table has a nest data path:
LOCATION is `/path/tablename/dt=20191113`. Data files are under `/path/tablename/dt=20191113/dt=20191113/xxx.parquet`.
By default, Spark can not recursively lookup the data files. So the UT I added can not return correct result. I open #26525 to enable recursively lookup in SQL.

But once recursively lookup enabled, the inferPartitioning will be disabled in current implementation. So this problem will be gone.

So should we trim out the inferring partition cols for Hive table? Or just throws a `CanNotConvertException` to handle it in upper layer?

**Comment by wangyum** [comment]
+1 for throws a `CanNotConvertException`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 6 above)
Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Reverted

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 7]
```diff
@@ -257,8 +257,20 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     }
     // The inferred schema may have different field names as the table schema, we should respect
     // it, but also respect the exprId in table relation output.
-    assert(result.output.length == relation.output.length &&
-      result.output.zip(relation.output).forall { case (a1, a2) => a1.dataType == a2.dataType })
+    if (result.output.length != relation.output.length) {
+      throw new HiveTableConvertException(
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: shall we just throw AnalysisException? It's not in the runtime stage yet.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 8]
```diff
@@ -358,4 +358,24 @@ class DataSourceWithHiveMetastoreCatalogSuite
...
+               |STORED AS PARQUET LOCATION '$baseDir'
+               |""".stripMargin)
+          val e = intercept[HiveTableConvertException](
+            spark.table("non_partition_table"), Seq()).getMessage
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: nit: what does `Seq()` mean here?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 9]
```diff
@@ -257,8 +257,20 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+        s"Converted table has ${result.output.length} columns, " +
+        s"but source Hive table has ${relation.output.length} columns. " +
+        s"Set ${HiveUtils.CONVERT_METASTORE_PARQUET.key} to false, " +
+        s"or recreate table ${relation.tableMeta.identifier} to workaround.")
+    }
+    if (!result.output.zip(relation.output).forall {
+          case (a1, a2) => a1.dataType.sameType(a2.dataType) }) {
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: This is a breaking change. We allow inferred schema may have different field names. `sameType` requires same field names.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala (hunk 8 above)
Comment (LantaoJin) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Ah, typo

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 9 above)
Comment (LantaoJin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Ok, I see. It restricts the field names by `l.name == r.name`. Sorry for that, I will revert this.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**